syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.registerallocation";
option java_outer_classname = "RegisterAllocationProto";
option objc_class_prefix = "REG";

package registerallocation;

// General Register Allocation Service 
service RegisterAllocation{

	// RPC function to get interference graphs
  rpc getGraphs(Path) returns (GraphList) {}
  // RPC function to send data for codegen
  rpc codeGen(Data) returns (RegisterProfileList) {}
 
}

message Path {
  string IRPath = 1;
  string SeedEmbPath=2;
}

message GraphList {
  bytes payload=1;
}


message RegisterProfileList {
  message RegisterProfile {
    float spillWeight=1;
    message interferenceOverlapsMap {
      uint32 interference = 1;
      repeated string overlaps = 2;
    }
    repeated interferenceOverlapsMap ioMap = 2;
  }
  repeated RegisterProfile regProf=1;
}

message Data {
  string message = 1;
  uint32 reg = 2;
  int64 payload = 3;
}

message Empty {

}



